### PR TITLE
Add github action to deploy to mvn central

### PR DESCRIPTION
**Pull request by veggiemonk**
# TL;DR

Add github action to deploy to mvn central

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Add the relevant configuration to deploy the artefacts to MavenCentral

## Tracking Issue

https://github.com/flyteorg/flyte/issues/2086

## Follow-up issue
_NA_

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -67,19 +67,6 @@
...
   </modules>
 
-  <licenses>
-    <license>
-      <name>Apache License, Version 2.0</name>
-      <url>https://www.apache.org/licenses/LICENSE-2.0</url>
-      <distribution>repo</distribution>
-    </license>
-  </licenses>
-  <scm>
```
Comment (narape) [inline pom.xml:77]: We removed this section but kept the this section in the github action
```
      - name: Release snapshot with Maven
        run: |
          mvn --batch-mode release:prepare  -DpreparationGoals=clean -DgenerateBackupPoms=false
          mvn --batch-mode release:perform  -DpreparationGoals=clean -Ddockerfile.push
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```

**Review by narape — COMMENTED** [review]
We need also modify the

File: pom.xml (hunk 1 above)
Comment (veggiemonk) [reply pom.xml:77]: re-added

File: pom.xml [hunk 2]
```diff
@@ -29,22 +29,24 @@
     <auto-service.version>1.0-rc5</auto-service.version>
     <auto-value.version>1.8.2</auto-value.version>
     <grpc.version>1.27.2</grpc.version>
-    <!-- must be aligned with netty version -->
-    <!-- see: https://github.com/grpc/grpc-java/blob/master/SECURITY.md#netty -->
+        <!-- must be aligned with netty version -->
+        <!-- see: https://github.com/grpc/grpc-java/blob/master/SECURITY.md#netty -->
```
Comment (narape) [inline pom.xml:33]: Let's keep the alignment that it had
```suggestion
    <!-- must be aligned with netty version -->
    <!-- see: https://github.com/grpc/grpc-java/blob/master/SECURITY.md#netty -->
```

File: .github/workflows/release.yaml [hunk 3]
```diff
@@ -34,10 +39,11 @@ jobs:
           registry: ghcr.io
           username: ${{ secrets.FLYTE_BOT_USERNAME }}
           password: ${{ secrets.FLYTE_BOT_PAT }}
-
-      - name: Release snapshot with Maven
+      
+      - name: Publish to MavenCentral
```
Comment (narape) [inline .github/workflows/release.yaml:43]: :+1: this is release, not snapshot, good catch

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -34,10 +39,11 @@ jobs:
...
         run: |
-          mvn --batch-mode release:prepare  -DpreparationGoals=clean -DgenerateBackupPoms=false
-          mvn --batch-mode release:perform  -DpreparationGoals=clean -Ddockerfile.push
+          mvn --batch-mode deploy -Pci-cd
```
Comment (narape) [inline .github/workflows/release.yaml:45]: [release:prepare](https://maven.apache.org/maven-release/maven-release-plugin/prepare-mojo.html) and then [release:perform](https://maven.apache.org/maven-release/maven-release-plugin/perform-mojo.html) deal with changing version and tagging the repository and then deploy.

I think that we want them back. Do you have a reason for removing it?

File: pom.xml [hunk 5]
```diff
@@ -29,22 +29,24 @@
...
     <junit.version>5.6.2</junit.version>
 
-    <!-- has to be one liner, or errorprone doesn't work -->
-    <error-prone-base-config><![CDATA[-Xplugin:ErrorProne -XepDisableWarningsInGeneratedCode -XepExcludedPaths:.*/generated-(test-)?sources/.*]]></error-prone-base-config>
-    <error-prone-additional-args>-Xep:AutoValueImmutableFields:OFF -Xep:Var:ERROR</error-prone-additional-args>
+        <!-- has to be one liner, or errorprone doesn't work -->
+    <error-prone-base-config>
+            <![CDATA[-Xplugin:ErrorProne -XepDisableWarningsInGeneratedCode -XepExcludedPaths:.*/generated-(test-)?sources/.*]]></error-prone-base-config>
+    <error-prone-additional-args>-Xep:AutoValueImmutableFields:OFF -Xep:Var:ERROR
+    </error-prone-additional-args>
```
Comment (narape) [inline pom.xml:49]: Let's roll back this change to keep indentation and alignment the same

File: pom.xml [hunk 6]
```diff
@@ -67,22 +69,9 @@
...
-  <scm>
-    <connection>scm:git:https://github.com/flyteorg/flytekit-java.git</connection>
-    <developerConnection>scm:git:https://github.com/flyteorg/flytekit-java.git</developerConnection>
-    <url>https://github.com/flyteorg/flytekit-java</url>
-    <tag>HEAD</tag>
-  </scm>
   <dependencyManagement>
     <dependencies>
-      <!-- modules -->
+            <!-- modules -->
```
Comment (narape) [inline pom.xml:74]: same here

File: pom.xml [hunk 7]
```diff
@@ -247,11 +236,36 @@
...
+    <developer>
+      <name>flytekit-java</name>
+      <email>dataex@spotify.com</email>
+      <organization>Flyte</organization>
+      <organizationUrl>https://flyte.org</organizationUrl>
+    </developer>
+  </developers>
   <distributionManagement>
+    <snapshotRepository>
+      <id>ossrh</id>
```
Comment (narape) [inline pom.xml:263]: It is not super clear in the [docs](https://maven.apache.org/pom.html#repository) if repository id must be unique among all other `<snapshotRepository>` or if it must be also unique with the other `<repository>`.

Can you verify if it's ok to repeat `ossrh`?

File: pom.xml [hunk 8]
```diff
@@ -247,11 +236,36 @@
...
+      <url>https://oss.sonatype.org/content/repositories/snapshots</url>
```
Comment (narape) [inline pom.xml:264]:
```suggestion
      <url>https://s01.oss.sonatype.org/content/repositories/snapshots/</url>
```
According to instructions [here](https://central.sonatype.org/publish/publish-guide/) this should be the new url

File: pom.xml [hunk 9]
```diff
@@ -247,11 +236,36 @@
...
+    </snapshotRepository>
     <repository>
-      <id>github</id>
-      <name>GitHub flyteorg flytekit-java Apache Maven Packages</name>
-      <url>https://maven.pkg.github.com/flyteorg/flytekit-java</url>
+      <id>ossrh</id>
+      <url>https://oss.sonatype.org/service/local/staging/deploy/maven2</url>
```
Comment (narape) [inline pom.xml:268]:
```suggestion
      <url>https://s01.oss.sonatype.org/service/local/staging/deploy/maven2/</url>
```
same instructions as before

File: pom.xml [hunk 10]
```diff
@@ -289,7 +303,7 @@
       </plugin>
       <plugin>
         <artifactId>maven-release-plugin</artifactId>
-        <!-- Downgrade to 2.4.1 if release fails -->
+                <!-- Downgrade to 2.4.1 if release fails -->
```
Comment (narape) [inline pom.xml:306]: it looks like this is an obsolete comment.

File: pom.xml [hunk 11]
```diff
@@ -311,20 +325,20 @@
             <showDeprecation>true</showDeprecation>
             <forceJavacCompilerUse>true</forceJavacCompilerUse>
             <failOnWarning>true</failOnWarning>
-            <!-- The semantics of this option are reversed, see MCOMPILER-209. -->
+                        <!-- The semantics of this option are reversed, see MCOMPILER-209. -->
```
Comment (narape) [inline pom.xml:328]: Please reverse this change, the same for the other ones

File: .github/workflows/release.yaml (hunk 4 above)
Comment (veggiemonk) [reply .github/workflows/release.yaml:45]: oh I thought they we to publish a container and other artefacts to github, I didn't understand why we would need a container for that. I'll add them back

File: pom.xml (hunk 7 above)
Comment (veggiemonk) [reply pom.xml:263]: It seems so as it is in their doc: https://central.sonatype.org/publish/publish-maven/

**Review by narape — APPROVED** [review]

**Comment by narape** [comment]
We broke master build because build depends on github repo that we just removed from the pom. Fix is coming in a follow up PR

**Pull request merged**